File: app/src/main/java/com/dmrandevu/gallery/media/VideoExporter.kt
package com.dmrandevu.gallery.media

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.media3.common.Effect
import androidx.media3.common.MediaItem
import androidx.media3.common.audio.AudioProcessor
import androidx.media3.common.util.UnstableApi
import androidx.media3.effect.OverlayEffect
import androidx.media3.transformer.Composition
import androidx.media3.transformer.EditedMediaItem
import androidx.media3.transformer.EditedMediaItemSequence
import androidx.media3.transformer.Effects
import androidx.media3.transformer.ExportException
import androidx.media3.transformer.ExportResult
import androidx.media3.transformer.ProgressHolder
import androidx.media3.transformer.Transformer
import com.dmrandevu.gallery.media.blur.BlurEffect
import com.dmrandevu.gallery.media.blur.BlurTimeline
import com.dmrandevu.gallery.media.blur.FaceFinder
import com.dmrandevu.gallery.media.blur.PlateFinder
import com.dmrandevu.gallery.media.blur.RegionFinder
import com.dmrandevu.gallery.media.blur.RegionScanner
import com.dmrandevu.gallery.media.censor.AudioCensor
import com.dmrandevu.gallery.media.censor.CensorPlan
import com.dmrandevu.gallery.media.censor.CensorWindow
import com.dmrandevu.gallery.media.censor.PatchingAudioProcessor
import com.dmrandevu.gallery.media.censor.ProfanityLexicon
import com.dmrandevu.gallery.media.watermark.WanderingWatermark
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import java.io.File
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Re-encodes a video with whatever [ExportOptions] asks for: faces mosaicked, licence plates
 * mosaicked, the account handle drifting across the picture, or any combination.
 *
 * Nothing here ever falls back to handing over the untouched video when something was asked for —
 * a silent unprotected export is exactly the failure these options exist to prevent, so a failed
 * pass throws instead.
 */
@UnstableApi
class VideoExporter(
    private val context: Context,
    private val audioCensor: AudioCensor
) {

    sealed interface Result {
        /**
         * [file] is the re-encoded video. [blurred] is what the scan actually covered, or null
         * when no blurring was asked for; it is worth handing back rather than re-running, since
         * a second scan would not agree with the first in every detail.
         */
        data class Exported(
            val file: File,
            val blurred: BlurTimeline?,
            /** The stretches that were beeped, or null when no censoring was asked for. */
            val censored: List<CensorWindow>? = null
        ) : Result

        /** Nothing to change, so the input was left alone — the caller delivers it as-is. */
        data object NothingToDo : Result
    }

    class ExportFailedException(message: String, cause: Throwable? = null) :
        Exception(message, cause)

    /**
     * Applies [options] to [input], writing to [output]. [onProgress] reports 0..100. Cancelling
     * the calling coroutine stops the export and removes [output].
     *
     * Every failure surfaces as [ExportFailedException] so callers have one thing to catch and no
     * way to mistake a broken export for a finished one.
     */
    suspend fun export(
        input: File,
        output: File,
        options: ExportOptions,
        onProgress: (Int) -> Unit
    ): Result {
        if (options.changesNothing) return Result.NothingToDo

        val finders = buildList<RegionFinder> {
            if (options.blurFaces) add(FaceFinder())
            if (options.blurPlates) {
                add(PlateFinder(context, PlateFinder.inputSizeFor(options.fastPlates)))
            }
        }

        // The two analysis passes share the run-up to the encode. Each takes the whole of it when
        // it is the only one asked for, so switching the censor off leaves the old numbers alone.
        val scanShare = if (finders.isEmpty()) 0 else if (options.censorAudio) SPLIT_SHARE else SCAN_SHARE
        val censorFloor = if (options.censorAudio) {
            if (finders.isEmpty()) SCAN_SHARE else SCAN_SHARE + SPLIT_SHARE
        } else {
            scanShare
        }

        val blurred = if (finders.isEmpty()) {
            null
        } else {
            wrapFailures("Scanning for faces and plates failed") {
                withContext(Dispatchers.Default) {
                    RegionScanner().scan(input, finders) { onProgress((it * scanShare).toInt()) }
                }
            }
        }

        val plan = if (!options.censorAudio) {
            null
        } else {
            wrapFailures("Censoring the audio failed") {
                audioCensor.analyze(
                    input,
                    tiersFor(options),
                    options.manualWindows,
                    byHand = options.censorByHand
                ) { percent ->
                    onProgress(scanShare + percent * (censorFloor - scanShare) / 100)
                }
            }
        }

        val effects = buildList<Effect> {
            // An empty timeline means nothing to cover; adding the effect anyway would re-encode
            // the whole video to change nothing.
            if (blurred != null && !blurred.isEmpty) add(BlurEffect(blurred))
            options.watermarkHandle?.let { add(OverlayEffect(listOf(WanderingWatermark(it)))) }
        }
        val processors = buildList<AudioProcessor> {
            if (plan != null && !plan.isEmpty) add(PatchingAudioProcessor(plan))
        }
        // Nothing found anywhere: the caller hands over the original, untouched and un-re-encoded.
        if (effects.isEmpty() && processors.isEmpty()) return Result.NothingToDo

        wrapFailures("Export failed") {
            runExport(input, output, effects, processors, censorFloor, onProgress)
        }
        // Raised here rather than mid-stream: media3 swallows what a processor throws while the
        // encode is running, and a mis-timed beep must fail the export, not ship.
        (processors.firstOrNull() as? PatchingAudioProcessor)?.failureOrNull()?.let {
            output.delete()
            throw ExportFailedException("Censoring the audio failed", it)
        }
        return Result.Exported(output, blurred, plan?.windows)
    }

    private fun tiersFor(options: ExportOptions) = buildSet {
        add(ProfanityLexicon.Tier.PROFANITY)
        if (options.censorInsults) add(ProfanityLexicon.Tier.INSULT)
    }

    private suspend fun <T> wrapFailures(message: String, block: suspend () -> T): T = try {
        block()
    } catch (e: CancellationException) {
        throw e
    } catch (e: ExportFailedException) {
        throw e
    } catch (e: Exception) {
        throw ExportFailedException(message, e)
    }

    /**
     * Whether this failure is the recognizer hearing swearing it could not place, rather than
     * anything being broken. The two need different words in front of the operator.
     */
    fun isUnplacedProfanity(failure: ExportFailedException): Boolean =
        (failure.cause as? AudioCensor.CensorFailedException)?.heardButUnplaced == true

    /**
     * Runs the Transformer export. Every touch of it — construction, start, progress, cancel —
     * has to happen on the main looper, which is why the whole body sits in [Dispatchers.Main].
     */
    private suspend fun runExport(
        input: File,
        output: File,
        effects: List<Effect>,
        audioProcessors: List<AudioProcessor>,
        progressFloor: Int,
        onProgress: (Int) -> Unit
    ) = withContext(Dispatchers.Main) {
        val editedMediaItem = EditedMediaItem.Builder(MediaItem.fromUri(Uri.fromFile(input)))
            .setEffects(Effects(audioProcessors, effects))
            .build()
        val sequence = EditedMediaItemSequence.Builder(editedMediaItem).build()
        val composition = Composition.Builder(sequence)
            // Copies the audio across untouched when nothing is censoring it. With a processor
            // in the list media3 transcodes regardless of this flag — for a single-item
            // composition it never consults it — but saying so keeps the intent on the record.
            .setTransmuxAudio(audioProcessors.isEmpty())
            // The shaders are ES 2.0 and cannot read HDR input. Tone mapping is a no-op for the
            // SDR videos Instagram actually delivers.
            .setHdrMode(Composition.HDR_MODE_TONE_MAP_HDR_TO_SDR_USING_OPEN_GL)
            .build()

        var poller: Job? = null
        try {
            suspendCancellableCoroutine { continuation ->
                val transformer = Transformer.Builder(context)
                    // Encode a portrait clip as portrait. Left off, media3 encodes it landscape
                    // and writes a rotation instead — correct, but players that ignore the
                    // rotation show the video on its side, and Instagram is not worth the gamble.
                    .setPortraitEncodingEnabled(true)
                    .addListener(object : Transformer.Listener {
                        override fun onCompleted(composition: Composition, result: ExportResult) {
                            if (continuation.isActive) continuation.resume(Unit)
                        }

                        override fun onError(
                            composition: Composition,
                            result: ExportResult,
                            exception: ExportException
                        ) {
                            if (continuation.isActive) {
                                continuation.resumeWithException(
                                    ExportFailedException("Transformer failed", exception)
                                )
                            }
                        }
                    })
                    .build()
                transformer.start(composition, output.absolutePath)

                poller = launch {
                    val holder = ProgressHolder()
                    while (true) {
                        delay(PROGRESS_POLL_MS)
                        if (transformer.getProgress(holder) == Transformer.PROGRESS_STATE_AVAILABLE) {
                            onProgress(progressFloor + holder.progress * (100 - progressFloor) / 100)
                        }
                    }
                }

                continuation.invokeOnCancellation {
                    // Cancellation arrives on whatever thread cancelled us, but Transformer only
                    // tolerates its own looper. This cannot go through `launch`: by now the scope
                    // is cancelled, so a new child would never run and the export would keep
                    // going with nobody left to stop it.
                    Handler(Looper.getMainLooper()).post {
                        runCatching { transformer.cancel() }
                        output.delete()
                    }
                }
            }
        } finally {
            poller?.cancel()
        }

        if (!output.exists() || output.length() == 0L) {
            throw ExportFailedException("Export produced no file")
        }
        onProgress(100)
    }

    private companion object {
        /** Share of the progress bar one analysis pass gets; the export takes the rest. */
        const val SCAN_SHARE = 35

        /** What each gets when the picture and the audio are both being analysed. */
        const val SPLIT_SHARE = 20
        const val PROGRESS_POLL_MS = 500L
    }
}

File: ios/DMRandevuGaleri/Media/VideoExporter.swift
import AVFoundation
import CoreImage
import Foundation

/// Re-encodes a video with whatever ``ExportOptions`` asks for: faces mosaicked, licence plates
/// mosaicked, the account handle drifting across the picture, or any combination.
///
/// Nothing here ever falls back to handing over the untouched video when something was asked for —
/// a silent unprotected export is exactly the failure these options exist to prevent, so a failed
/// pass throws instead.
final class VideoExporter {

    /// Finds the swearing and renders the audio that replaces it.
    let audioCensor: AudioCensor

    init(audioCensor: AudioCensor) {
        self.audioCensor = audioCensor
    }

    enum Result {
        /// `url` is the re-encoded video. `blurred` is what the scan actually covered, or nil when
        /// no blurring was asked for; it is worth handing back rather than re-running, since a
        /// second scan would not agree with the first in every detail.
        case exported(url: URL, blurred: BlurTimeline?, censored: [CensorWindow]? = nil)

        /// Nothing to change, so the input was left alone — the caller delivers it as-is.
        case nothingToDo
    }

    struct ExportFailedError: LocalizedError {
        let message: String
        let underlying: Error?

        var errorDescription: String? {
            underlying.map { "\(message): \($0.localizedDescription)" } ?? message
        }
    }

    /// Shared by every plate scan: a Core Image context is expensive to build and safe to reuse.
    private let context: CIContext = {
        if let device = MTLCreateSystemDefaultDevice() {
            return CIContext(mtlDevice: device)
        }
        return CIContext()
    }()

    /// Applies `options` to `input`, writing to `output`. `onProgress` reports 0…100. Cancelling
    /// the calling task stops the export and removes `output`.
    ///
    /// Every failure surfaces as ``ExportFailedError`` so callers have one thing to catch and no
    /// way to mistake a broken export for a finished one.
    func export(
        input: URL,
        output: URL,
        options: ExportOptions,
        onProgress: @escaping (Int) -> Void
    ) async throws -> Result {
        if options.changesNothing { return .nothingToDo }

        var finders: [RegionFinder] = []
        if options.blurFaces { finders.append(FaceFinder()) }
        if options.blurPlates {
            guard let plates = PlateFinder(fast: options.fastPlates, context: context) else {
                throw ExportFailedError(message: "Plate detector missing", underlying: nil)
            }
            finders.append(plates)
        }

        // The two analysis passes share the run-up to the encode. Each takes the whole of it when
        // it is the only one asked for, so switching the censor off leaves the old numbers alone.
        let scanShare = finders.isEmpty
            ? 0
            : (options.censorAudio ? Self.splitShare : Self.scanShare)
        let censorFloor = options.censorAudio
            ? (finders.isEmpty ? Self.scanShare : Self.scanShare + Self.splitShare)
            : scanShare

        var blurred: BlurTimeline?
        if !finders.isEmpty {
            blurred = try await wrapping("Scanning for faces and plates failed") {
                try await RegionScanner().scan(url: input, finders: finders) { fraction in
                    onProgress(Int(fraction * Double(scanShare)))
                }
            }
        }

        var plan: CensorPlan?
        if options.censorAudio {
            var tiers: Set<ProfanityLexicon.Tier> = [.profanity]
            if options.censorInsults { tiers.insert(.insult) }
            plan = try await wrapping("Censoring the audio failed") {
                try await audioCensor.analyze(input: input, tiers: tiers) { percent in
                    onProgress(scanShare + percent * (censorFloor - scanShare) / 100)
                }
            }
        }

        // An empty timeline means nothing to cover; running the pass anyway would re-encode the
        // whole video to change nothing.
        let coverage = (blurred?.isEmpty == false) ? blurred : nil
        let censoring = (plan?.isEmpty == false) ? plan : nil
        if coverage == nil && options.watermarkHandle == nil && censoring == nil {
            return .nothingToDo
        }

        let source = AVURLAsset(url: input)
        let composition = try await wrapping("Building the filter pass failed") {
            try await VideoFilterPipeline.composition(
                for: source,
                blur: coverage,
                watermark: options.watermarkHandle
            )
        }

        // With the audio replaced the export runs over a composition rather than the file, since
        // this platform gives no way to rewrite samples on their way through.
        var asset: AVAsset = source
        var censoredAudio: URL?
        if let censoring {
            let audioURL = output.deletingLastPathComponent()
                .appendingPathComponent("censored-audio.caf")
            censoredAudio = audioURL
            asset = try await wrapping("Censoring the audio failed") {
                try await CensoredAudioTrack.write(plan: censoring, from: input, to: audioURL)
                return try await Self.compose(video: source, audio: audioURL)
            }
        }
        defer { censoredAudio.map { try? FileManager.default.removeItem(at: $0) } }

        try await wrapping("Export failed") {
            try await run(
                asset: asset,
                composition: composition,
                to: output,
                floor: censorFloor,
                onProgress: onProgress
            )
        }
        return .exported(url: output, blurred: blurred, censored: plan?.windows)
    }

    /// The original picture with the censored audio in place of its own.
    private static func compose(video: AVURLAsset, audio: URL) async throws -> AVAsset {
        let composition = AVMutableComposition()
        let range = CMTimeRange(start: .zero, duration: try await video.load(.duration))

        if let track = try await video.loadTracks(withMediaType: .video).first,
           let into = composition.addMutableTrack(
               withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid
           ) {
            try into.insertTimeRange(range, of: track, at: .zero)
            // Carried across explicitly: without it a portrait clip composes as landscape.
            into.preferredTransform = try await track.load(.preferredTransform)
        }

        let censored = AVURLAsset(url: audio)
        if let track = try await censored.loadTracks(withMediaType: .audio).first,
           let into = composition.addMutableTrack(
               withMediaType: .audio, preferredTrackID: kCMPersistentTrackID_Invalid
           ) {
            let audioRange = CMTimeRange(
                start: .zero,
                duration: min(range.duration, try await censored.load(.duration))
            )
            try into.insertTimeRange(audioRange, of: track, at: .zero)
        }
        return composition
    }

    /// Runs the encode. Audio is re-encoded rather than copied across, which the export presets
    /// give no way around; at these presets that is inaudible, and Instagram re-encodes again on
    /// upload regardless.
    private func run(
        asset: AVAsset,
        // Nil when only the audio is being censored: the picture then passes through untouched
        // and there is no filter pass to install.
        composition: AVMutableVideoComposition?,
        to output: URL,
        floor: Int,
        onProgress: @escaping (Int) -> Void
    ) async throws {
        // The session refuses to write over anything, including a file a cancelled run left.
        try? FileManager.default.removeItem(at: output)

        guard let session = AVAssetExportSession(
            asset: asset,
            presetName: AVAssetExportPresetHighestQuality
        ) else {
            throw ExportFailedError(message: "No encoder for this video", underlying: nil)
        }
        session.videoComposition = composition

        let monitor = Task {
            for await state in session.states(updateInterval: Self.progressInterval) {
                guard case .exporting(let progress) = state else { continue }
                let share = Double(100 - floor) * progress.fractionCompleted
                onProgress(floor + Int(share))
            }
        }
        defer { monitor.cancel() }

        do {
            try await session.export(to: output, as: .mp4)
        } catch {
            try? FileManager.default.removeItem(at: output)
            throw error
        }

        let size = (try? FileManager.default.attributesOfItem(atPath: output.path)[.size] as? Int) ?? 0
        guard FileManager.default.fileExists(atPath: output.path), size > 0 else {
            throw ExportFailedError(message: "Export produced no file", underlying: nil)
        }
        onProgress(100)
    }

    /// Turns anything that is not a cancellation into the one error type callers handle.
    private func wrapping<T>(_ message: String, _ work: () async throws -> T) async throws -> T {
        do {
            return try await work()
        } catch is CancellationError {
            throw CancellationError()
        } catch let failure as ExportFailedError {
            throw failure
        } catch {
            throw ExportFailedError(message: message, underlying: error)
        }
    }

    /// Share of the progress bar the scan gets; the export takes the rest.
    private static let scanShare = 35

    /// What each analysis pass gets when the picture and the audio are both being looked at.
    private static let splitShare = 20
    private static let progressInterval: TimeInterval = 0.5
}
